x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichShardMultiSearchAction.java

```diff
@@ -0,0 +1,268 @@
+            super(NAME, threadPool, clusterService, transportService, actionFilters, indexNameExpressionResolver,
+                Request::new, ThreadPool.Names.SEARCH);
+            this.indicesService = indicesService;
+        }
+
+        @Override
+        protected Writeable.Reader<MultiSearchResponse> getResponseReader() {
+            return MultiSearchResponse::new;
+        }
+
+        @Override
+        protected boolean resolveIndex(Request request) {
+            return true;
+        }
+
+        @Override
+        protected ShardsIterator shards(ClusterState state, InternalRequest request) {
+            String index = request.concreteIndex();
+            IndexRoutingTable indexRouting = state.routingTable().index(index);
+            int numShards = indexRouting.shards().size();
+            if (numShards != 1) {
+                throw new IllegalStateException("index [" + index + "] should have 1 shard, but has " + numShards + " shards");
+            }
+            return indexRouting.shard(0).shardsRandomIt();
+        }
+
+        @Override
+        protected MultiSearchResponse shardOperation(Request request, ShardId shardId) throws IOException {
+            final long nowInMillis = System.currentTimeMillis();
+            final IndexService indexService = indicesService.indexService(shardId.getIndex());
+            final IndexShard indexShard = indicesService.getShardOrNull(shardId);
+            try (Engine.Searcher searcher = indexShard.acquireSearcher("enrich_msearch")) {
+                final FieldsVisitor visitor = new FieldsVisitor(true);
+                final QueryShardContext context =
+                    indexService.newQueryShardContext(shardId.id(), searcher.getIndexReader(), () -> nowInMillis, null);
+                final MapperService mapperService = context.getMapperService();
+                final Text typeText = mapperService.documentMapper().typeText();
+
+                final MultiSearchResponse.Item[] items = new MultiSearchResponse.Item[request.multiSearchRequest.requests().size()];
+                for (int i = 0; i < request.multiSearchRequest.requests().size(); i++) {
+                    final SearchSourceBuilder searchSourceBuilder = request.multiSearchRequest.requests().get(i).source();
+
+                    final QueryBuilder queryBuilder = searchSourceBuilder.query();
+                    final int from = searchSourceBuilder.from();
+                    final int size = searchSourceBuilder.size();
+                    final FetchSourceContext fetchSourceContext = searchSourceBuilder.fetchSource();
+
+                    final Query luceneQuery = queryBuilder.rewrite(context).toQuery(context);
+                    final int n = from + size;
+                    final TopDocs topDocs = searcher.search(luceneQuery, n, new Sort(SortField.FIELD_DOC));
+
+                    final SearchHit[] hits = new SearchHit[topDocs.scoreDocs.length];
```

Comment:
do we need multiple hits ? For exact_match it will always be 1 or 0.

Reply:
The idea is that this internal api is shared by other implementations as well.
Currently `ExactMatchProcessor` checks whether either no or exactly one result is returned. But a `exact` or `geo` policy would not do that, but it should be possible from to reuse this internal api.